scipy/sparse/tests/test_csr.py

```diff
@@ -167,3 +170,17 @@ def test_csr_hstack_int64():
     X_hs_32 = hstack([X_1, X_3], format="csr")
     assert X_hs_32.indices.dtype == np.int32
     assert X_hs_32.indices.max() == max_int32 - 1
+
+@pytest.mark.parametrize("cls", [csr_matrix, csr_array, csc_matrix, csc_array])
+def test_mixed_index_dtype_int_indexing(cls):
+    # https://github.com/scipy/scipy/issues/20182
+    rng = np.random.default_rng(0)
+    base_mtx = cls(sparse.random(50, 50, random_state=rng, density=0.1))
+    indptr_64bit = base_mtx.copy()
+    indices_64bit = base_mtx.copy()
+    indptr_64bit.indptr = base_mtx.indptr.astype(np.int64)
+    indptr_64bit.indices = base_mtx.indices.astype(np.int64)
+
+    indices = [([2, 3, 4], slice(None)), (slice(None), [2, 3, 4])]
+    for idx, mtx in product(indices, [base_mtx, indptr_64bit, indices_64bit]):
+        np.testing.assert_array_equal(mtx[idx].toarray(), base_mtx[idx].toarray())
```

Comment:
It looks like CJ already approved, but I did notice that when I revert your source change this regression test still passes.

Something like this, which borrows from your original issue reproducer, fails before and passes after the source patch:

```diff
--- a/scipy/sparse/tests/test_csr.py
+++ b/scipy/sparse/tests/test_csr.py
@@ -183,4 +183,6 @@ def test_mixed_index_dtype_int_indexing(cls):
 
     indices = [([2, 3, 4], slice(None)), (slice(None), [2, 3, 4])]
     for idx, mtx in product(indices, [base_mtx, indptr_64bit, indices_64bit]):
-        np.testing.assert_array_equal(mtx[idx].toarray(), base_mtx[idx].toarray())
\ No newline at end of file
+        np.testing.assert_array_equal(mtx[idx].toarray(), base_mtx[idx].toarray())
+        base_mtx.indptr = base_mtx.indptr.astype(np.int64)
+        base_mtx[[1, 2], :]
```

Any chance we could do something like that? I'm not a `sparse` regular, so maybe it can be cleaner than that.

Reply:
Thanks for the catch! I've corrected this, though am a little unsure in what exactly is different.